### PR TITLE
Refs #20577 -- Deferred filtering of prefetched related querysets by reverse m2o relation.

### PR DESCRIPTION
**Pull request by bobbyrenwick**
https://github.com/django/django/pull/11916, released in v3.1, introduced the idea of deferring filtering of a QS when prefetching related many to many relations, but the many to one case was left out. This adds the same idea to the many to one case.

I've never really stepped into the django codebase before so please let me know if there's anything else that needs doing. Tests, tickets etc.

Thanks!

### CODE REVIEW

**Comment by felixxm** [comment]
@bobbyrenwick Thanks for this patch :+1: There is no need for a new ticket, we can refer to the original.

Test is required, see `test_filter_deferred` in `tests/prefetch_related/tests.py` for an inspiration.

File: tests/prefetch_related/tests.py [hunk 1]
```diff
@@ -306,6 +306,28 @@ def test_filter_deferred(self):
...
+    def test_foreignkey_forward_filter_deferred(self):
+        add_q = Query.add_q
+        with mock.patch.object(
+            Query,
+            'add_q',
+            autospec=True,
+            side_effect=lambda self, q: add_q(self, q),
+        ) as add_q_mock:
+            list(Book.objects.prefetch_related('first_book'))
+            self.assertEqual(add_q_mock.call_count, 1)
```
Comment (felixxm) [inline tests/prefetch_related/tests.py:318]: `Book` model doesn't have the `first_book` attribute, moreover this test is not related with fix. I think we can chop it.

File: tests/prefetch_related/tests.py [hunk 2]
```diff
@@ -306,6 +306,28 @@ def test_filter_deferred(self):
...
+    def test_foreignkey_reverse_filter_deferred(self):
+        add_q = Query.add_q
+        with mock.patch.object(
+            Query,
+            'add_q',
+            autospec=True,
+            side_effect=lambda self, q: add_q(self, q),
+        ) as add_q_mock:
+            list(Book.objects.prefetch_related('first_time_authors'))
+            self.assertEqual(add_q_mock.call_count, 1)
```
Comment (felixxm) [inline tests/prefetch_related/tests.py:329]: We can use `self.subTest()` and add test for a m2o relation to the `test_filter_deferred`, e.g.
```python
    def test_filter_deferred(self):
        """
        Related filtering of prefetched querysets is deferred until necessary.
        """
        add_q = Query.add_q
        for relation in ['authors', 'first_time_authors']:
            with self.subTest(relation=relation):
                with mock.patch.object(
                    Query,
                    'add_q',
                    autospec=True,
                    side_effect=lambda self, q: add_q(self, q),
                ) as add_q_mock:
                    list(Book.objects.prefetch_related(relation))
                    self.assertEqual(add_q_mock.call_count, 1)
```

File: tests/prefetch_related/tests.py (hunk 1 above)
Comment (bobbyrenwick) [reply tests/prefetch_related/tests.py]: Hey, I meant to use `Author` here but got lost in a copy/paste. I have fixed the reference in d8e75e154a4e441fb0b02c0850f15234f2d3d1b9 and think it's still relevant?

File: tests/prefetch_related/tests.py (hunk 2 above)
Comment (bobbyrenwick) [reply tests/prefetch_related/tests.py:329]: Thanks for the heads up - done in 776c6261751d7b7c30b76fe8dcc67ffb1c244cae

File: tests/prefetch_related/tests.py (hunk 1 above)
Comment (felixxm) [reply tests/prefetch_related/tests.py]: It's not related with this patch, because it works without it. Also, as far as I'm aware this is not even related with an optimization added in  681f7e2b13846def9eff3d26ce1371d05fa90c4c, because it works without it. I don't think we need this test. Please remove it.

Comment (bobbyrenwick) [reply tests/prefetch_related/tests.py]: :+1: removed in 31cfd3f1f2dca6b0759230692c33489b5ff61434

**Comment by felixxm** [comment]
@bobbyrenwick Thanks for updates :+1: Welcome aboard :boat:
